### PR TITLE
fix(aws): AllowLaunchAtomicOperation - Add retries when resolving AMI

### PR DESCRIPTION
**Pull request by christopherthielen**

### CODE REVIEW

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy [hunk 1]
```diff
@@ -61,28 +62,13 @@ class AllowLaunchAtomicOperation implements AtomicOperation<ResolvedAmiResult> {
     def sourceAmazonEC2 = amazonClientProvider.getAmazonEC2(description.credentials, description.region, true)
     def targetAmazonEC2 = amazonClientProvider.getAmazonEC2(targetCredentials, description.region, true)
 
+    def (ResolvedAmiResult resolvedAmi, ResolvedAmiLocation amiLocation) = new RetrySupport().retry({ ->
```
Comment (christopherthielen) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy]: this style OK?

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy [hunk 2]
```diff
@@ -61,28 +62,13 @@ class AllowLaunchAtomicOperation implements AtomicOperation<ResolvedAmiResult> {
...
+      resolveAmi(targetCredentials, sourceAmazonEC2, targetAmazonEC2)
+    }, 5, 1000, false)
```
Comment (christopherthielen) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy]: these numbers OK? I copied from elsewhere.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy [hunk 3]
```diff
@@ -61,28 +62,13 @@ class AllowLaunchAtomicOperation implements AtomicOperation<ResolvedAmiResult> {
...
-        task.updateStatus BASE_PHASE, "AMI appears to be from a private third-party, which is permitted on this target account: skipping allow launch"
-        return resolvedAmi
-      }
-    }
-
-    if (!resolvedAmi) {
-      throw new IllegalArgumentException("unable to resolve AMI imageId from '$description.amiName': If this is a private AMI owned by a third-party, you will need to contact them to share the AMI to your desired account(s)")
+    if (amiLocation == ResolvedAmiLocation.THIRD_PARTY) {
+      task.updateStatus BASE_PHASE, "AMI appears to be from a private third-party, which is permitted on this target account: skipping allow launch"
+      return resolvedAmi
```
Comment (christopherthielen) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy]: early exit for 3rd party

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy [hunk 4]
```diff
@@ -94,17 +80,17 @@ class AllowLaunchAtomicOperation implements AtomicOperation<ResolvedAmiResult> {
     // If the AMI was created/owned by a different account, switch to using that for modifying the image
     if (resolvedAmi.ownerId != sourceCredentials.accountId) {
       if (resolvedAmi.getRegion())
-      sourceCredentials = accountCredentialsProvider.all.find { accountCredentials ->
-        accountCredentials instanceof NetflixAmazonCredentials &&
-          ((AmazonCredentials) accountCredentials).accountId == resolvedAmi.ownerId
-      } as NetflixAmazonCredentials
+        sourceCredentials = accountCredentialsProvider.all.find { accountCredentials ->
```
Comment (christopherthielen) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy]: unintentional reformat, I will revert

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy [hunk 5]
```diff
@@ -153,6 +139,37 @@ class AllowLaunchAtomicOperation implements AtomicOperation<ResolvedAmiResult> {
     resolvedAmi
   }
 
+  private static enum ResolvedAmiLocation {
+    TARGET, SOURCE, THIRD_PARTY
+  }
+
+  private Tuple2<ResolvedAmiResult, ResolvedAmiLocation> resolveAmi(
```
Comment (christopherthielen) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy]: Tuple2 ok? seemed like a decent way to return 2 values with type information and destructure the result

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy (hunk 4 above)
Comment (christopherthielen) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy]: actually, this lil guy is missing braces.  I will retain the reformat and add braces

Comment (cfieber) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy]: :+1:

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy (hunk 5 above)
Comment (cfieber) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy]: Yeah seems legit

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy (hunk 2 above)
Comment (cfieber) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy]: should be ok - 5 retries 1sec apart. I don't think we want to retry too much longer than that in here, if it turns out this isn't enough maybe the next step is to look at moving things up to orca.

**Pull request merged**